File: app/services/ai/stakeholder_identifier.rb
# frozen_string_literal: true

module Ai
  class StakeholderIdentifier < BaseAnalyzer
    SYSTEM_PROMPT = <<~PROMPT
      You are an expert at identifying stakeholders affected by product insights.
      Analyze the insight and determine which user segments, teams, or groups
      should be engaged or considered.

      For each stakeholder segment:
      1. Create a clear name for the segment
      2. Classify the type: user_segment, internal_team, customer_tier, use_case_group, or geographic_region
      3. Write a description of this stakeholder group
      4. Estimate the size of this segment
      5. Rate engagement priority (0-5, where 5 is critical)
      6. Suggest an engagement strategy
      7. List characteristics that define this segment

      Respond in JSON format:
      {
        "stakeholders": [
          {
            "name": "Segment name",
            "segment_type": "user_segment|internal_team|customer_tier|use_case_group|geographic_region",
            "description": "Who this segment is",
            "size_estimate": 500,
            "engagement_priority": 4,
            "engagement_strategy": "How to engage this group",
            "characteristics": ["Characteristic 1", "Characteristic 2"],
            "impact_level": 3,
            "impact_description": "How this insight affects them"
          }
        ]
      }

      Guidelines:
      - Identify 2-5 relevant stakeholder segments
      - Include both internal teams (engineering, support, sales) and user segments
      - Higher priority for segments most affected by the insight
      - Be specific about engagement strategies
    PROMPT

    def identify(insight)
      result = analyze_insight(insight)

      if result[:error]
        Rails.logger.error("Stakeholder identification failed for insight ##{insight.id}: #{result[:error]}")
        return { stakeholders: [], created: 0, error: result[:error] }
      end

      created = create_stakeholders_from_result(result, insight)

      { stakeholders: created, created: created.count }
    end

    def identify_batch(insights)
      all_stakeholders = []

      insights.each do |insight|
        result = identify(insight)
        all_stakeholders.concat(result[:stakeholders]) unless result[:error]
        rate_limit_sleep
      end

      { stakeholders: all_stakeholders.uniq, created: all_stakeholders.count }
    end

    private

    def analyze_insight(insight)
      prompt = build_insight_prompt(insight)
      call_claude(prompt, system_prompt: SYSTEM_PROMPT)
    end

    def build_insight_prompt(insight)
      parts = ["Identify stakeholders for the following insight:\n"]
      parts << "Title: #{insight.title}"
      parts << "Description: #{insight.description}"
      parts << "Type: #{insight.insight_type}"
      parts << "Severity: #{insight.severity}"
      parts << "Affected Users: #{insight.affected_users_count}"

      if insight.feedbacks.any?
        parts << "\nSample Feedback Sources:"
        insight.feedbacks.limit(5).each do |f|
          parts << "- #{f.source}: #{f.author_name || 'Anonymous'}"
        end
      end

      parts.join("\n")
    end

    def create_stakeholders_from_result(result, insight)
      created_stakeholders = []

      stakeholders_data = result[:stakeholders] || []
      stakeholders_data.each do |stakeholder_data|
        stakeholder = find_or_create_stakeholder(stakeholder_data)
        next unless stakeholder

        link_stakeholder_to_insight(stakeholder, insight, stakeholder_data)
        created_stakeholders << stakeholder
      end

      created_stakeholders
    end

    def find_or_create_stakeholder(data)
      existing = StakeholderSegment.find_by("LOWER(name) = ?", data[:name].downcase)
      return update_stakeholder(existing, data) if existing

      StakeholderSegment.create!(
        name: data[:name],
        segment_type: data[:segment_type],
        description: data[:description],
        size_estimate: data[:size_estimate] || 0,
        engagement_priority: data[:engagement_priority] || 0,
        engagement_strategy: data[:engagement_strategy],
        characteristics: data[:characteristics] || []
      )
    rescue ActiveRecord::RecordInvalid => e
      Rails.logger.error("Failed to create stakeholder: #{e.message}")
      nil
    end

    def update_stakeholder(stakeholder, data)
      stakeholder.update!(
        size_estimate: [stakeholder.size_estimate, data[:size_estimate] || 0].max,
        engagement_priority: [stakeholder.engagement_priority, data[:engagement_priority] || 0].max
      )
      stakeholder
    end

    def link_stakeholder_to_insight(stakeholder, insight, data)
      InsightStakeholder.find_or_create_by!(
        insight: insight,
        stakeholder_segment: stakeholder
      ) do |is|
        is.impact_level = data[:impact_level] || 0
        is.impact_description = data[:impact_description]
      end
    end
  end
end

File: apps/worker/src/ai/stakeholder-identifier.ts
/**
 * AI stakeholder identification — ports Rails StakeholderIdentifier.
 * Identifies affected user segments per insight.
 */
import { eq, sql } from "drizzle-orm";
import type { Database } from "@customer-pulse/db/client";
import { insights, stakeholderSegments, insightStakeholders } from "@customer-pulse/db/client";
import { callClaudeJson } from "./call-claude.js";

interface IdentifiedStakeholder {
  name: string;
  segment_type: string;
  description: string;
  size_estimate: number;
  engagement_priority: number;
  engagement_strategy: string;
  characteristics: string[];
  impact_level: number;
  impact_description: string;
}

const SEGMENT_TYPE_MAP: Record<string, number> = { user_segment: 0, internal_team: 1, customer_tier: 2, use_case_group: 3, geographic_region: 4 };

const SYSTEM_PROMPT = `You are a stakeholder analyst. Given an insight about customer feedback, identify 2-5 stakeholder segments that are affected.

For each stakeholder, return:
- "name": segment name (< 60 chars)
- "segment_type": one of "user_segment", "internal_team", "customer_tier", "use_case_group", "geographic_region"
- "description": 1-2 sentence description
- "size_estimate": estimated number of people in this segment
- "engagement_priority": 0-5 how urgently this segment should be engaged (5=most urgent)
- "engagement_strategy": recommended approach for this segment
- "characteristics": array of key traits
- "impact_level": 0-4 how severely this insight impacts this segment
- "impact_description": brief description of the impact

Return a JSON array. Respond with ONLY the JSON array, no markdown fences.`;

export async function identifyStakeholders(
  db: Database,
  insightId: number,
): Promise<number> {
  const [insight] = await db.select().from(insights).where(eq(insights.id, insightId)).limit(1);
  if (!insight) return 0;

  const typeNames: Record<number, string> = { 0: "problem", 1: "opportunity", 2: "trend", 3: "risk", 4: "user_need" };
  const result = await callClaudeJson<IdentifiedStakeholder[]>({
    system: SYSTEM_PROMPT,
    user: `Insight: "${insight.title}" (${typeNames[insight.insightType]})\n\n${insight.description}`,
    maxTokens: 2048,
  });

  if (!result || !Array.isArray(result)) return 0;

  const now = new Date();
  let created = 0;

  for (const item of result) {
    // Find or create stakeholder segment by name
    const [existing] = await db
      .select({ id: stakeholderSegments.id })
      .from(stakeholderSegments)
      .where(sql`lower(${stakeholderSegments.name}) = ${(item.name ?? "").toLowerCase()} AND ${stakeholderSegments.projectId} = ${insight.projectId}`)
      .limit(1);

    let segmentId: number;
    if (existing) {
      segmentId = existing.id;
    } else {
      const [newSeg] = await db.insert(stakeholderSegments).values({
        projectId: insight.projectId,
        name: item.name ?? "Unknown Segment",
        segmentType: SEGMENT_TYPE_MAP[item.segment_type] ?? 0,
        description: item.description ?? null,
        sizeEstimate: item.size_estimate ?? 0,
        engagementPriority: item.engagement_priority ?? 0,
        engagementStrategy: item.engagement_strategy ?? null,
        characteristics: (item.characteristics ?? []) as unknown[],
        metadata: {},
        createdAt: now,
        updatedAt: now,
      }).returning({ id: stakeholderSegments.id });
      if (!newSeg) continue;
      segmentId = newSeg.id;
      created++;
    }

    // Link to insight
    try {
      await db.insert(insightStakeholders).values({
        insightId,
        stakeholderSegmentId: segmentId,
        impactLevel: item.impact_level ?? 2,
        impactDescription: item.impact_description ?? null,
        createdAt: now,
        updatedAt: now,
      });
    } catch { /* duplicate — ignore */ }
  }

  return created;
}
